src/esrs/aggregate.rs

```diff
@@ -20,69 +20,76 @@ pub trait Identifier {
-/// persisting those events in the store. Be careful when implementing this trait, as you will be responsible for
-/// threading AggregateState/Commands/Events correctly.  For example, a bad implementation could result in an AggregateState
-/// that could not be replicated on load.
+/// The Aggregate trait is responsible for validating commands, mapping commands to events, and applying
+/// events onto the state.
 ///
-/// Unless you need to perform side effects as part of your command handling/verification you should implement the
-/// safer `Aggregate` trait instead.
-#[async_trait]
-pub trait AggregateManager: Identifier {
+/// An Aggregate should be able to derive its own state from nothing but its initial configuration, and its
+/// event stream. Applying the same events, in the same order, to the same aggregate, should always yield an
+/// identical aggregate state.
+///
+/// This trait is purposefully _synchronous_. If you are implementing this trait, your aggregate
+/// should not have any side effects. If you need additional information to handle commands correctly, then
+/// consider looking up that information and placing it in the command.
+pub trait Aggregate: Identifier {
     type State: Default + Clone + Debug + Send + Sync;
     type Command: Send + Sync;
     type Event: Serialize + DeserializeOwned + Send + Sync;
     type Error: Send + Sync;
 
-    /// Returns the event store, configured for the aggregate
-    fn event_store(&self) -> &(dyn EventStore<Self::Event, Self::Error> + Send + Sync);
+    /// Validates a command against the current aggregate state. The aggregate must be able to handle the command
+    /// if validation succeeds.
+    fn validate_command(state: &AggregateState<Self::State>, cmd: &Self::Command) -> Result<(), Self::Error>;
 
-    /// This function applies the event onto the aggregate and returns a new one, updated with the event data
-    fn apply_event(id: &Uuid, state: Self::State, event: &StoreEvent<Self::Event>) -> Self::State;
+    /// Handles a validated command, and emits events. This must always succeed if the command was validated
+    /// correctly.
+    ///
+    /// Passing a non-validated command is allowed to panic.
+    fn handle_command(state: &AggregateState<Self::State>, cmd: Self::Command) -> Vec<Self::Event>;
 
-    /// Validation should reject any command is inconsistent with the current aggregate state, or would result
-    /// in one or more events that could not be applied onto the aggregate state.
-    fn validate_command(aggregate_state: &AggregateState<Self::State>, cmd: &Self::Command) -> Result<(), Self::Error>;
+    /// Updates the aggregate state using the new event. This assumes that the event can be correctly applied
+    /// to the state.
+    ///
+    /// If this is not the case, this function is allowed to panic.
+    fn apply_event(state: Self::State, payload: &Self::Event) -> Self::State;
```

Comment:
I'm not sure about this signature. Maybe state should be an `AggregateState<Self::State>` and payload should be `&StoreEvent<Self::Event>` in order to use some data like `occurred_on`, `sequence_number` or the event id?

Reply:
The Aggregate only sees the business logic, it should not see the implementation details of `StoreEvent<_>` imho